docs/specs/lcb-traits.md

```diff
@@ -0,0 +1,302 @@
+        -- Perform formatting on T
+    end handler
+
+In this case, the compiler would identify that `M.FormatAsString<T>()` has a
+name that matches a generic handler declaration, and that it conforms (all
+generic parameters are specialised and all non-generic parameters are of the
+correct types).  Subsequent calls to `FormatAsString()` from *anywhere in the
+program* would then be able to access `M.FormatAsString<T>()` if the first
+argument of `FormatAsString()` conforms to `T`.
+
+The main drawback here is that Builder is optionally-typed, but it would be
+good to ensure that it's possible to write code where all types are specified
+*and can be verified* at compile time.  This doesn't satisfy the requirement.
+
+For example, it would be convenient to write a `Log()` handler that can log
+any type that can also be formatted.  However, with the generic handlers scheme
+described above, `Log()` would have to take arguments as `any` and defer
+type-checking to runtime.
+
+## Traits
+
+Traits are collections of methods that can be used to extend the functionality
+of a type.
+
+Traits have no storage associated with them.  Storage is provided by
+implementing the trait for a specific storage type.
+
+### Defining traits
+
+    public trait type CanFormatAsString
+        method type FormatAsString(in me) as string
+    end trait
+
+A trait definition is a block containing a list zero or more `method`
+definitions or `method type` declarations.  `method type` declarations take the
+same form as existing `handler type` declarations.  `method` definitions take
+the same form as existing `handler` declarations.
+
+    public trait type CanFormatAsString
+        method FormatAsString(in me) as string
+            return "<unknown>"
+        end method
+    end trait
+
+`method` definitions allow a trait's definition to include default
+implementations of the trait's methods.  Default implementations can call other
+methods of the trait or perform any operations valid for the generic `Self`
+type.
+
+Two additional symbols are defined within a trait definition:
+
+1. `MyType` is always equivalent to the type of the value that the method is
+   being called for.
+2. `me` must always be the first parameter a `method` definition or
+   `method type` declaration; it is equivalent to writing `me as MyType`.  The
+   `me` parameter must always be `in` or `inout`.
+
+### Implementing traits
+
+    public implementation of CanFormatAsString for T
+        method FormatAsString(in me) as string
+            variable tFormatted as string
+            -- Format me parameter
+            return tFormatted
+        end method
+    end implementation
+
+The `me` and `MyType` symbols are also defined inside implementation blocks.
+
+In order for an `implementation of U for T` to be valid, the following criteria
+must be fulfilled:
+
+1. Either `U` or `T` must be local, i.e. defined in the current module.  This
```

Comment:
> Wouldn't that create an entirely separate trait though? i.e. Any consumers anywhere of the trait implemented in Faff for FooBar, would be relying on the trait defined in Faff.

Yes, just like if there'd been a useful handler function of `FooBar` copied into the original module. It'd be an entirely different handler, and any consumers of the handler implemented in `Faff` would be relying on it to be defined in `Faff`. But that's okay because they'd just use the earlier version of the `FooBar` module API which doesn't have the colliding symbol, right?

> I'm assuming that the implementation would essentially need to add each trait implementation to the list implemented by a given storage type - otherwise generic 'format list as string operations wouldn't work...

Sometimes it's useful to have private implementations.  For example, you might want to implement `CanFormatAsString` for a `public type T` for debugging purposes internally to your module but also want to prevent users of your module from from formatting `T` (maybe because your debugging output is a. hideous and b. exposes implementation details you'd rather not exhibit).